### PR TITLE
virt-api, admitter: Refactor & move network validation admitters to the network package

### PR DESCRIPTION
**Pull request by EdDev**
### What this PR does

Validation admitters related to network spec are moved under the network package in order to decouple them from the other admitters and allow the sig-network to properly own them.

In addition, work on simplifying the validation code and its tests has been included in this effort.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: pkg/network/admitter/slirp.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	k8sfield "k8s.io/apimachinery/pkg/util/validation/field"
+
+	"kubevirt.io/kubevirt/pkg/network/vmispec"
+
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+
+	v1 "kubevirt.io/api/core/v1"
+)
+
+func ValidateSlirpBinding(field *k8sfield.Path, spec *v1.VirtualMachineInstanceSpec, config *virtconfig.ClusterConfig) (causes []metav1.StatusCause) {
```
Comment (ormergi) [inline pkg/network/admitter/slirp.go]: I am missing who call this function

**Review by ormergi — COMMENTED** [review]
Hi Edy, thanks for the PR

I the network related addmiters code is moved to `pkg/network`, how about keep it under `pkg/virt-api/webhooks/validating-webhook/admitters/` but in new dir for network: `pkg/virt-api/webhooks/validating-webhook/admitters/network`?

Since the code is tightly related to the webhook and slirp/ iface APIs, I think it make more sense leaving it closer to to the consuming package.
It may also encourage other sigs who doesn't aggregate all related code under pkg/ move their related content in similar why to what I suggest.

**Comment by EdDev** [comment]
> Hi Edy, thanks for the PR
> 
> I the network related addmiters code is moved to `pkg/network`, how about keep it under `pkg/virt-api/webhooks/validating-webhook/admitters/` but in new dir for network: `pkg/virt-api/webhooks/validating-webhook/admitters/network`?
> 
> Since the code is tightly related to the webhook and slirp/ iface APIs, I think it make more sense leaving it closer to to the consuming package. It may also encourage other sigs who doesn't aggregate all related code under pkg/ move their related content in similar why to what I suggest.

This decision was already made when the `admitter` package was introduced under the `network` root package. What change that needs this to be reexamined?

Most of the current `network` sub-packages are used by many components in Kubevirt, I see no special thing about the admission ones.

File: pkg/network/admitter/slirp.go (hunk 1 above)
Comment (EdDev) [reply pkg/network/admitter/slirp.go]: Great catch.
I am not sure if it got lost in the rebases, but it is something I need to make sure it is covered.

Done

When all network validations will move out, I will add a check that they are  called in the main flow.

**Comment by EdDev** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/ac9d8b9d19bffc687687ec11bfa739b4817e29ef..99a546d33a567f19b48af93f8d445ff37b60020a): Fix a bug detected during the review.

**Comment by ormergi** [comment]
> This decision was already made when the `admitter` package was introduced under the `network` root package. What change that needs this to be reexamined?
> 

I was under the impression this PR introduce `addmiter` package, I missed that one.

**Review by ormergi — APPROVED** [review]

**Comment by EdDev** [comment]
change: Rebase

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go [hunk 2]
```diff
@@ -1354,45 +1354,6 @@ var _ = Describe("Validating VMICreate Admitter", func() {
 			Expect(causes).To(HaveLen(1))
 			Expect(causes[0].Field).To(Equal("fake.networks[0]"))
 		})
-		It("should reject networks with a multus network source and slirp interface", func() {
-			enableSlirpInterface()
```
Comment (AlonaKaplan) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go:1369]: I think this method should now be removed from the file, isn't it?
I found some usages of it, but it is on tests checking ports. It seems other binding can be used instead.

File: pkg/network/admitter/slirp.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		if ifaceSpec.Slirp == nil {
+			continue
+		}
+		net := vmispec.LookupNetworkByName(spec.Networks, ifaceSpec.Name)
+		if net == nil {
+			continue
+		}
+
+		if net.Pod == nil {
+			causes = append(causes, metav1.StatusCause{
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/network/admitter/slirp.go:49]: It was not supported per what I understand.

File: pkg/network/admitter/slirp.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+				Type:    metav1.CauseTypeFieldValueInvalid,
+				Message: "Slirp interface only implemented with pod network",
+				Field:   field.Child("domain", "devices", "interfaces").Index(idx).Child("name").String(),
+			})
+		} else if !config.IsSlirpInterfaceEnabled() {
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/network/admitter/slirp.go]: If we do not lookup the network, we do not know if it exists or not.
I do not know if there is a correct answer to this.
E.g. if there is no network, we ignore this error and let some other validation to kick in.
Also, if it is not a pod network source, we error on that and not on the slirp being disabled.

Somehow I think the current check is good, checking the slirp enablement only if all other things are good.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go (hunk 2 above)
Comment (EdDev) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go:1369]: I prefer to remove it once I move everything out and not now. It reduces the number of changes needed.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go [hunk 5]
```diff
@@ -284,6 +284,7 @@ func validateNetworksMatchInterfaces(field *k8sfield.Path, spec *v1.VirtualMachi
 
 		networkData, networkExists := networkNameMap[iface.Name]
 
+		causes = append(causes, netadmitter.ValidateSlirpBinding(field, spec, config)...)
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go]: Done

**Comment by EdDev** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/e24ff7a1985836ca6cf4d549bc5d1d6a96248542..108a131f0d8041b3f4a981fd3a7e06815e406f2c): Answered a review comment to move out the slirp validation outside the for-loop.

**Comment by EdDev** [comment]
change: Rebase

File: pkg/network/admitter/BUILD.bazel [hunk 6]
```diff
@@ -15,11 +15,15 @@ go_library(
 
 go_test(
     name = "go_default_test",
-    srcs = ["admit_test.go"],
+    srcs = [
```
Comment (RamLavi) [inline pkg/network/admitter/BUILD.bazel:21]: nit: please consider moving this commit to before the test is moved.

**Pull request merged**